plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/TableSnapshot.java

```diff
@@ -259,7 +280,87 @@ private CheckpointEntryIterator getCheckpointTransactionLogEntries(
                 checkpointRowStatisticsWritingEnabled,
                 domainCompactionThreshold,
                 partitionConstraint,
-                addStatsMinMaxColumnFilter);
+                addStatsMinMaxColumnFilter));
+    }
+
+    private Stream<DeltaLakeTransactionLogEntry> getV2CheckpointTransactionLogEntriesFrom(
+            ConnectorSession session,
+            Set<CheckpointEntryIterator.EntryType> entryTypes,
+            Optional<MetadataEntry> metadataEntry,
+            Optional<ProtocolEntry> protocolEntry,
+            CheckpointSchemaManager checkpointSchemaManager,
+            TypeManager typeManager,
+            FileFormatDataSourceStats stats,
+            LastCheckpoint checkpoint,
+            TrinoInputFile checkpointFile,
+            TupleDomain<DeltaLakeColumnHandle> partitionConstraint,
+            Optional<Predicate<String>> addStatsMinMaxColumnFilter,
+            TrinoFileSystem fileSystem,
+            long fileSize)
+    {
+        Optional<List<DeltaLakeTransactionLogEntry>> checkpointLogEntries;
+        if (checkpointFile.location().fileName().endsWith(".json")) {
+            try {
+                checkpointLogEntries = getEntriesFromJson(checkpoint.getVersion(), checkpointFile);
+            }
+            catch (IOException e) {
+                throw new TrinoException(DELTA_LAKE_FILESYSTEM_ERROR, format("Unexpected IO exception occurred while reading the entries of the file: %s for the table %s", checkpoint, table), e);
+            }
+        }
+        else if (checkpointFile.location().fileName().endsWith(".parquet")) {
+            CheckpointEntryIterator parquetCheckpoint = new CheckpointEntryIterator(
+                    fileSystem.newInputFile(checkpointFile.location()),
+                    session,
+                    fileSize,
+                    checkpointSchemaManager,
+                    typeManager,
+                    ImmutableSet.<CheckpointEntryIterator.EntryType>builder()
+                            .addAll(entryTypes)
+                            .add(SIDECAR)
+                            .build(),
+                    metadataEntry,
+                    protocolEntry,
+                    stats,
+                    parquetReaderOptions,
+                    checkpointRowStatisticsWritingEnabled,
+                    domainCompactionThreshold,
+                    partitionConstraint,
+                    addStatsMinMaxColumnFilter);
+            checkpointLogEntries = Optional.of(ImmutableList.copyOf(parquetCheckpoint));
+        }
+        else {
+            throw new IllegalArgumentException("Unsupported v2 checkpoint file format: " + checkpointFile.location());
+        }
+
+        if (checkpointLogEntries.isEmpty()) {
+            return Stream.empty();
+        }
+        Stream<DeltaLakeTransactionLogEntry> transactionLogs = Stream.empty();
```

Comment:
This algorithm seems to be solvable with streams & flatmap (in a similar fashion to what is being done in https://github.com/trinodb/trino/blob/b90fe09f4f28ee1f2cc6a74283b311972242d673/plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/TableSnapshot.java#L200-L217 )